src/Mumble.proto

```diff
@@ -20,6 +20,35 @@ message Version {
 	optional string os_version = 4;
 }
 
+// Send after the version packet, indicate the protocol preference of the client.
+// - This message is not required for text-only clients. It can be send before or after
+//   the Authenticate message. However, it has be sent prior to a client attempts to send
+//   or receive any audio data.
+// - As reply, the server will send an Capabilities message as well, with only one item in
+//   the list, indicating the protocol it determines to use.
+// - In the middle of a session, a client can re-perform this negotiation by send a valid
+//   Capabilities packet again. A server may also request the client to use another Capabilities
+//   with the protocol it'd like to use inside.
+message Capabilities {
+	enum VoiceProtocol {
+		UNDEFINED = 0;  // Placeholder used by the server to indicate it doesn't support any protocol the client has provided.
+		CUSTOM = 1;
+		UDP_OCB2 = 2;
+	}
+
+	// A list of supported protocols of the client sent to the server, sort by preference.
+	// If the client wants to include a custom protocol that is not listed in VoiceProtocol,
+	// he can use CUSTOM as placeholders. Server will look up custom_protocols and map CUSTOM
+	// with the protocol names inside custom_protocols in turn.
+	repeated VoiceProtocol supported_protocols = 1;
+	// Protocol names used to fill the CUSTOM placeholders appear in supported_protocols.
+	repeated string custom_protocols = 2;
```

Comment:
This is not the solution that we have discussed in #4248. Having only one field for this does make more sense than having 2 separate fields...

Reply:
I tried. That one was ugly and hard to manage. You will have really long generated enum name like 
```
MumbleProto::Capabilities_VoiceProtocol_VoiceTransport_Official_UDP_OCB2
```

I think this is purely my personal's aesthetical choice, and I stand with it.

Reply:
According to Johni's proposal I'd expect the enum name to be `MumbleProto::VoiceTransport::Official::UDP_OCB2`. Besides: If you think the name is ugly, just typedef it like so:
```
using VoiceProtocol = MumbleProto::VoiceTransport::Official
```
Then you can simply use `VoiceProtocol::UDP_OCB2` everywhere in the code.

Or if it does indeed generate this ugly long name, I think it's not even an enum class. We should change that by creating a new enum class inside our code with the respective values. In order to not have code-duplication we should simply assign the enum values based on the ones in the proto file:
```
enum class VoiceProtocol {
    UDP_OCB2 = MumbleProto::Capabilities_VoiceProtocol_VoiceTransport_Official_UDP_OCB2
}
```

Reply:
The way you've proposed to avoid duplication is impressive and I will adapt it for sure.

However, I'm against Johni's suggestion.
In another word, I just don't want to see nested messages. 
That's something IMO looks _structured_ but unnecessarily confusing. Instead, I would better not abuse the flexibilities of protobuf and build structured monsters with it.

Maybe we can consult more people to see their preferences.

Reply:
What's the problem with nested messages? It's basically the same as having a class definined within a class in c++. That's perfectly normal if that inner class is not needed anywhere on the outside...

IMO it is also more clear to have one message field for one purpose and not 2 fields for 1 purpose. That's why I like John's suggestion.

As there seems to be some more room for discussion though, let's switch to #4248 for it :)
Maybe you could elaborate on why exactly you think a nested message is confusing? :thinking: